lib/spack/spack/spec.py

```diff
@@ -4845,26 +4845,37 @@ def reconstruct_virtuals_on_edges(spec):
         spec: spec on which we want to reconstruct virtuals
     """
     # Collect all possible virtuals
-    possible_virtuals = set()
+    possible_virtuals = collections.defaultdict(set)
     for node in spec.traverse():
+        key = node.dag_hash()
         try:
-            possible_virtuals.update(
-                {x for x in node.package.dependencies if spack.repo.PATH.is_virtual(x)}
-            )
+            node_pkg = node.package
         except Exception as e:
-            warnings.warn(f"cannot reconstruct virtual dependencies on package {node.name}: {e}")
+            warnings.warn(f"cannot reconstruct virtual dependencies on {node.name}: {e}")
             continue
 
-    # Assume all incoming edges to provider are marked with virtuals=
-    for vspec in possible_virtuals:
+        for name, when_deps in node_pkg.dependencies_by_name(when=True).items():
+            if not spack.repo.PATH.is_virtual(name):
+                continue
+
+            if any(node.satisfies(x) for x in when_deps):
+                possible_virtuals[key].add(name)
+
+    for edge in spec.traverse_edges(cover="edges"):
+        if edge.parent is None:
+            continue
```

Comment:
```suggestion
    for edge in spec.traverse_edges(cover="edges", root=False):
```
or is this to deal with mypy?

Reply:
maybe skip if `key not in provided_virtuals`? and just use a normal dict instead of defaultdict?

Reply:
Actually, I don't get why you're moving the parent related stuff into a loop of its own, but leave the child bit here.

Would it make more sense to do a single loop over edges only and check whether you've already computed what the parent needs and what the child provides as you go? That's probably shorter.

Reply:
The last implementation iterates once over all edges